// @appium/typedoc-plugin-appium/build/lib/model/reflection/commands.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommandsReflection = void 0;
const typedoc_1 = require("typedoc");
const kind_1 = require("./kind");
/**
 * A reflection containing data about commands and/or execute methods.
 *
 * Methods may be invoked directly by Handlebars templates.
 */
class CommandsReflection extends typedoc_1.DeclarationReflection {
    constructor(name, parent, { routeMap, execMethodDataSet }) {
        super(name, kind_1.AppiumPluginReflectionKind.COMMANDS, parent);
        this.parent = parent;
        this.routeMap = routeMap;
        this.execMethodDataSet = execMethodDataSet;
    }
    /**
     * Returns `true` if there are any "execute commands" in this set.
     *
     * Used by templates
     */
    get hasExecuteMethod() {
        return Boolean(this.execMethodCount);
    }
    /**
     * Returns `true` if there are any "regular" commands in this set.
     *
     * Used by templates
     */
    get hasRoute() {
        return Boolean(this.routeCount);
    }
    /**
     * Returns number of routes ("commands") in this in this data
     */
    get routeCount() {
        return this.routeMap.size;
    }
    /**
     * Returns number of execute methods in this data
     */
    get execMethodCount() {
        return this.execMethodDataSet.size;
    }
}
exports.CommandsReflection = CommandsReflection;
