// @appium/docutils/build/lib/logger.js
"use strict";
/**
 * It's a logger.
 *
 * Since this is a CLI app only, it doesn't necessarily make sense to consume `@appium/support`'s logger.
 *
 * @module
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const figures_1 = __importDefault(require("figures"));
const log_symbols_1 = __importDefault(require("log-symbols"));
const chalk_1 = __importDefault(require("chalk"));
const consola_1 = __importStar(require("consola"));
const constants_1 = require("./constants");
const lodash_1 = __importDefault(require("lodash"));
/**
 * This is a reporter for `consola` which uses some extra/custom icons and colors.
 *
 * @privateRemarks
 * I did not like that the default `FancyReport` logs errors in _green_ without any sort of icon.
 * Both `log-symbols` and `consola` consume `chalk`, so we do too. `consola` also depends on `figures`.
 */
class DocutilsReporter extends consola_1.FancyReporter {
    /**
     * Type guard to check if a log type has a color
     * @param type A log type
     */
    static hasTypeColor(type) {
        return type in DocutilsReporter.TYPE_COLOR_MAP;
    }
    /**
     * Type guard to check if a log level has a color
     * @param level A log level
     */
    static hasLevelColor(level) {
        return level in DocutilsReporter.LEVEL_COLORS;
    }
    /**
     * Type guard to check if a log type has an icon
     * @param type A log type
     */
    static hasTypeIcon(type) {
        return type in DocutilsReporter.TYPE_ICONS;
    }
    /**
     * Prefixes the logging output with colors and symbols, depending on contents of `logObj`.
     * @param logObj Consola's log object
     * @param isBadge {@linkcode FancyReporter} uses this; I think it depends on the terminal width
     * @returns
     */
    formatType(logObj, isBadge) {
        const { TYPE_COLOR_MAP, LEVEL_COLORS, TYPE_ICONS, hasTypeColor, hasLevelColor, hasTypeIcon, DEFAULT_COLOR, } = DocutilsReporter;
        let typeColor;
        if (hasTypeColor(logObj.type)) {
            typeColor = TYPE_COLOR_MAP[logObj.type];
        }
        else if (hasLevelColor(logObj.level)) {
            typeColor = LEVEL_COLORS[logObj.level];
        }
        else {
            typeColor = ((this.options.secondaryColor ?? DEFAULT_COLOR));
        }
        if (isBadge) {
            return chalk_1.default[('bg' + lodash_1.default.capitalize(typeColor))].black(` ${lodash_1.default.toUpper(logObj.type)}`);
        }
        const type = hasTypeIcon(logObj.type) ? TYPE_ICONS[logObj.type] : logObj.type;
        return type ? chalk_1.default[typeColor](type) : '';
    }
}
/**
 * Mapping of log types (the name of the logging method called) to chalk fg colors
 */
DocutilsReporter.TYPE_COLOR_MAP = {
    info: 'cyan',
    success: 'green',
    error: 'red',
    warn: 'yellow',
};
/**
 * Mapping of log levels to chalk fg colors
 */
DocutilsReporter.LEVEL_COLORS = {
    0: 'red',
    1: 'yellow',
    2: 'white',
    3: 'green',
};
/**
 * Mapping of log types to icons/symbols
 */
DocutilsReporter.TYPE_ICONS = {
    info: log_symbols_1.default.info,
    success: log_symbols_1.default.success,
    error: log_symbols_1.default.error,
    warn: log_symbols_1.default.warning,
    debug: (0, figures_1.default)('›'),
    trace: (0, figures_1.default)('›'),
};
/**
 * Default color to use if we can't find a color for the log type or level
 */
DocutilsReporter.DEFAULT_COLOR = 'grey';
/**
 * The global log level
 *
 * "Global" inasmuch as any logger created from the root logger will use this level.
 */
let globalLevel = constants_1.LogLevelMap[constants_1.DEFAULT_LOG_LEVEL];
/**
 * The logger from which all loggers are created.  This one uses a unique tag and our custom reporter.
 */
const rootLogger = createLogProxy(consola_1.default.create({ defaults: { tag: 'docutils' }, reporters: [new DocutilsReporter()] }));
/**
 * @summary Creates a log-level-propagating proxy for a {@linkcode Consola} logger.
 * @description
 * Alright.  So when you create a new logger via {@linkcode Consola.create}, it's basically a clone
 * of its parent with a new set of options.
 *
 * If we change the log level of the root logger (which we do: see `cli/index.ts`), we may (almost
 * certainly) have
 * child loggers which: a) have already been created and b) have inherited the old/default log level
 * from the root logger. We don't _want_ that (though this is likely a reasonable use case) for our
 * purposes.
 *
 * The implementation below solves the problem by maintaining its own singleton log level value, and
 * intercepts the `level` property of any logger created from the root logger.
 *
 * There are other ways to go about this which may be better, but this seemed pretty straightforward.
 */
function createLogProxy(logger) {
    return new Proxy(logger, {
        get(target, prop, receiver) {
            if (prop === 'level') {
                return globalLevel;
            }
            if (prop === 'create') {
                const create = Reflect.get(target, prop, receiver);
                return (opts) => createLogProxy(create.call(receiver, opts));
            }
            if (prop === '_defaults') {
                const defaults = Reflect.get(target, prop, receiver);
                return { ...defaults, level: globalLevel };
            }
            return Reflect.get(target, prop, receiver);
        },
        set(target, prop, value, receiver) {
            if (prop === 'level') {
                globalLevel = value;
                return true;
            }
            return Reflect.set(target, prop, value, receiver);
        },
    });
}
/**
 * The proxied root logger
 * @see {createLogProxy}
 */
exports.default = rootLogger;
// these are just type-sanity checks
DocutilsReporter.TYPE_COLOR_MAP;
DocutilsReporter.LEVEL_COLORS;
DocutilsReporter.TYPE_ICONS;
